---
title: "GSE120575"
author: "Lingyi"
date: "6/3/2019"
output: html_document
---
Identify individual markers associated with response and lack of response. 
Method: 'Differential expression analysis' part
Data: gene expression values of 6,350 CD8+ T cells (cell ID's found in Table S2)
Truth: Table S2

# Select data
```{r}
knitr::opts_chunk$set(
  cache = TRUE,
  warning = FALSE,
  message = FALSE,
  cache.lazy = FALSE)

dataPath   <- "./GSE120575Data/"
outPath <- "./output/"

library(data.table)
# library(cowplot)     
# library(RColorBrewer)
# library(limma)
library(dplyr)
library(readxl)
```

```{r}
data <- fread("./GSE120575Data/GSE120575_Sade_Feldman_melanoma_single_cells_TPM_GEO.txt", sep="\t", header = TRUE, fill = TRUE)
data.frame <- as.data.frame(data)
rname <- data.frame[, 1]
dfAll <- data.frame[, -1]
row.names(dfAll) <- rname
dfAll$V16293 <- NULL
```

```{r}
anno <- read_excel("./GSE120575Data/GSE120575_patient_ID_single_cells.xlsx", sheet="GSE120575_patient_ID_single_cells")

Responder <- filter(anno,  Post_onTreatment == "Responder") 
Responder <- unique(Responder$patinet_ID)

NonResponder <- filter(anno,  Post_onTreatment == "Non-responder") 
NonResponder <- unique(NonResponder$patinet_ID)
```


```{r}
# need to simplify 
Resindex <- which(dfAll[1, ] == "Pre_P1" | dfAll[1, ] == "Post_P1" | dfAll[1, ] == "Post_P4" | dfAll[1, ] == "Post_P5_2" | dfAll[1, ] == "Pre_P7" | dfAll[1, ] == "Post_P7" | dfAll[1, ] == "Pre_P8" | dfAll[1, ] == "Post_P8" | dfAll[1, ] == "Post_P17" | dfAll[1, ] == "Post_P19" | dfAll[1, ] == "Post_P21" | dfAll[1, ] == "Pre_P24" | dfAll[1, ] == "Pre_P26" | dfAll[1, ] == "Pre_P28" | dfAll[1, ] == "Pre_P29" | dfAll[1, ] == "Pre_P33"| dfAll[1, ] == "Pre_P35" )
```


```{r}
# (3) cells with an average expression of housekeeping genes (Table S6)
# divide to response and nonresponse 
S6 <- read_excel("./GSE120575Data/tableS6.xlsx", sheet="Protein coding genes-methods")
S6.genes <- unlist(S6, use.names=FALSE)
df <- dfAll[S6.genes, ]
dfRes <- df[, Resindex]
dfNon <- df[, -Resindex]
```


```{r}
# prepare the data: gene expression values of 6,350 CD8+ T cells
S2 <- read_excel("./GSE120575Data/tableS2.xlsx", sheet="Cluster annotation-Fig2A-B")
S2.Tcell <- S2[, 1]
S2.Tcell <- unlist(S2.Tcell, use.names=FALSE)
df.Tcell <- df[, S2.Tcell]
```

```{r}
Rcommon <- intersect(colnames(dfRes), colnames(df.Tcell))  
dfR <- select(dfRes, Rcommon)
NRcommon <- intersect(colnames(dfNon), colnames(df.Tcell))  
dfNR <- select(dfNon, NRcommon)
```

```{r}
clusters <- unlist(S1.cluster[,1],  use.names=FALSE)
```

# Analysis
## Classify R & NR
```{r}
FMatrix.RNR <- function(gene){
  A <- length(which(as.numeric(dfR[gene,]) > 2))
  B <- length(which(as.numeric(dfNR[gene,])>2))
  C <- length(which(as.numeric(dfR[gene,]) <= 2))
  D <- length(which(as.numeric(dfNR[gene,]) <= 2))
  mean_G1 <- mean(as.numeric(dfR[gene,]))
  mean_G2 <- mean(as.numeric(dfNR[gene,]))
  perR <- (A / (A+C))
  perNR <- (B / (B+D))
  FMatrix <- matrix(c(A, B, C, D), nrow = 2, dimnames = list(c("G1","G2"), c("log2(TPM+1)>2","log2(TPM+1)<=2")))
  p <- fisher.test(FMatrix)[["p.value"]]
  p
  compute <- log2(mean_G1/mean_G2)
  out <- data.frame("p_value" = p, "mean_exp_in_R" = mean_G1, "mean_exp_in_NR" = mean_G2, 
                    "per_in_R" = perR, "per_in_NR" = perNR, "log2_R_NR_" = log2(mean_G1/mean_G2))
  out
  # print(FMatrix)
  # print(sprintf("Mean expression G1: %f", mean_G1))
  # print(sprintf("Mean expression G2: %f", mean_G2))
  # print(sprintf("Percentage in R: %f", perN))
  # print(sprintf("Percentage in NR: %f", perNR))
  # print(sprintf("log2(R/NR): %f",log2(mean_G1/mean_G2)))
  # print(p.adjust(p, method = "bonferroni", n = length(p)))
}
```

```{r}
FMatrix.RNR("CD38")
```

```{r}
test <- unlist(S6, use.names = FALSE)
length(test)
```

```{r}
# around two hours to run this part
genesPvalue <- sapply(test, FMatrix.RNR)
genesPvalue <- as.data.frame(t(genesPvalue))
# l <- list()
# i <- 1
# for (gene in test){
#   l[i] <- FMatrix.function(gene)
#   i <- i + 1
# }
# 
# genesPvalue <- data.frame(matrix(unlist(l), nrow=length(l), byrow=T))
# rownames(genesPvalue) <- test
```

```{r}
# sort 
pvalueSort <- genesPvalue
pvalueSort <- as.data.frame(lapply(pvalueSort, unlist))
#geneidSort <- geneID[order(pvalueSort$p_value)]
pvalueSort <- pvalueSort[order(pvalueSort$p_value), ]
pvalueSort <- as.data.frame(pvalueSort)
#rownames(pvalueSort) <- geneidSort
```

```{r}
write.csv(pvalueSort, file = "./GSE120575Data/pvalueSort.csv")
```


## Classify CD8_G & CD8_B
```{r}
# two clusters G1: CD8_G, G2: CD8_B
S2 <- read_excel("./GSE120575Data/tableS2.xlsx", sheet="Cluster annotation-Fig2A-B")
CD8_G <- filter(S2, Cluster == "CD8_G") # 3327
CD8_G <- unlist(CD8_G[, 1],  use.names=FALSE)
df.CD8_G <- select(df.Tcell, CD8_G)

CD8_B <- filter(S2, Cluster == "CD8_B") # 3032 
CD8_B <- unlist(CD8_B[, 1],  use.names=FALSE)
df.CD8_B <- select(df.Tcell, CD8_B)
```
  
```{r}
FMatrix.BG <- function(gene){
  A <- length(which(as.numeric(df.CD8_B[gene,]) > 2))
  B <- length(which(as.numeric(df.CD8_G[gene,])>2))
  C <- length(which(as.numeric(df.CD8_B[gene,]) <= 2))
  D <- length(which(as.numeric(df.CD8_G[gene,]) <= 2))
  mean_G1 <- mean(as.numeric(df.CD8_B[gene,]))
  mean_G2 <- mean(as.numeric(df.CD8_G[gene,]))
  FMatrix <- matrix(c(A, B, C, D), nrow = 2, dimnames = list(c("G1","G2"), c("log2(TPM+1)>2","log2(TPM+1)<=2")))
  p <- fisher.test(FMatrix)[["p.value"]]
  p
  out <- data.frame("p_value" = p, "mean_exp_in_G1" = mean_G1, "mean_exp_in_G2" = mean_G2)
  out
  # print(FMatrix)
  # print(sprintf("Mean expression G1: %f", mean_G1))
  # print(sprintf("Mean expression G2: %f", mean_G2))
  # print(sprintf("Percentage in R: %f", perN))
  # print(sprintf("Percentage in NR: %f", perNR))
  # print(sprintf("log2(R/NR): %f",log2(mean_G1/mean_G2)))
  # print(p.adjust(p, method = "bonferroni", n = length(p)))
}
```

```{r}
FMatrix.BG("CD38")
```

## Classify G1 & G2 

```{r}
# dont have to run again
library(writexl)
# S1.cluster <- read_excel("./GSE120575Data/tableS1.xlsx", sheet="Cluster annotation-Fig1B-C")
# S1.cluster[, 1] <- colnames(dfAll)
write_xlsx(S1.cluster, "./GSE120575Data/tableS1.xlsx")
```

```{r}
# Find G1 
S1.cluster <- read_excel("./GSE120575Data/cluster.xlsx", sheet="Sheet1")

cluster_1 <- filter(S1.cluster, Cluster_number == 1) # 1455
cluster_1 <- unlist(cluster_1[, 1],  use.names=FALSE)
#common <- intersect(cluster_1, colnames(df))  
G1 <- select(df, cluster_1) 

# Find G2
cluster_2 <- filter(S1.cluster, Cluster_number == 2) # 305 
cluster_2 <- unlist(cluster_2[, 1],  use.names=FALSE)
#common <- intersect(cluster_2, colnames(df))  
G2 <- select(dfAll, cluster_2)
```


